backend/wallet/services/offchain/payment.py

```diff
@@ -0,0 +1,234 @@
+import dataclasses
+import logging
+from datetime import datetime
+from typing import Optional
+
+from diem_utils.types.currencies import DiemCurrency
+from offchain import (
+    jws,
+)
+from offchain.types import (
+    new_get_info_request,
+    GetInfoCommandResponse,
+    new_init_charge_command,
+    new_init_auth_command,
+    InitChargePaymentResponse,
+)
+from wallet import storage
+from wallet.services.offchain import utils
+from wallet.services.offchain.utils import generate_my_address
+from wallet.storage.models import Payment as PaymentModel
+from wallet.storage.payment import save_payment
+from wallet.types import TransactionType, TransactionStatus
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(frozen=True)
+class PaymentDetails:
+    vasp_address: str
+    reference_id: str
+    merchant_name: str
+    action: str
+    currency: str
+    amount: int
+    expiration: int
+
+
+class P2MGeneralError(Exception):
+    pass
+
+
+class PaymentNotFoundError(Exception):
+    pass
+
+
+def get_payment_details(account_id, reference_id: str, vasp_address: str):
+    payment_model = storage.get_payment_details(reference_id)
+
+    if payment_model is None:
+        my_address = utils.generate_my_address(account_id)
+
+        try:
+            command_response_object = utils.offchain_client().send_request(
+                request_sender_address=my_address,
+                counterparty_account_id=vasp_address,
+                request_bytes=jws.serialize(
+                    new_get_info_request(reference_id=reference_id, cid=reference_id),
+                    utils.compliance_private_key().sign,
+                ),
+            )
+        except Exception as e:
+            error = P2MGeneralError(e)
+            logger.error(error)
+            raise error
+
+        if (
+            command_response_object.result
+            and type(command_response_object.result) is GetInfoCommandResponse
```

Comment:
This is awfully not pythonic. The pythonic way is to use the object and catch the error of the needed attributes are not there. Or at least do not check the type – the the existence of the payment_info attribute. In general, if you check types in Python, something is wrong.